cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp

```diff
@@ -0,0 +1,171 @@
+using namespace Ice;
+using namespace IceInternal;
+
+namespace
+{
+    class HeartbeatTimerTask final : public IceUtil::TimerTask
+    {
+    public:
+        HeartbeatTimerTask(const ConnectionIPtr& connection) : _connection(connection) {}
+
+        void runTimerTask() final
+        {
+            if (auto connection = _connection.lock())
+            {
+                connection->sendHeartbeat();
+            }
+            // else nothing to do, the connection is already gone.
+        }
+
+    private:
+        const std::weak_ptr<ConnectionI> _connection;
+    };
+
+    class IdleCheckTimerTask final : public IceUtil::TimerTask, public std::enable_shared_from_this<IdleCheckTimerTask>
+    {
+    public:
+        IdleCheckTimerTask(const ConnectionIPtr& connection, const chrono::milliseconds& idleTimeout)
+            : _connection(connection),
+              _idleTimeout(idleTimeout)
+        {
+        }
+
+        void runTimerTask() final
+        {
+            if (auto connection = _connection.lock())
+            {
+                connection->idleCheck(shared_from_this(), _idleTimeout);
+            }
+            // else nothing to do, the connection is already gone.
+        }
+
+    private:
+        const std::weak_ptr<ConnectionI> _connection;
+        const chrono::milliseconds _idleTimeout;
+    };
+}
+
+void
+IdleTimeoutTransceiverDecorator::decoratorInit(const ConnectionIPtr& connection)
+{
+    _heartbeatTimerTask = make_shared<HeartbeatTimerTask>(connection);
+    if (_enableIdleCheck)
+    {
+        _idleCheckTimerTask = make_shared<IdleCheckTimerTask>(connection, _idleTimeout);
+    }
+}
+
+SocketOperation
+IdleTimeoutTransceiverDecorator::initialize(Buffer& readBuffer, Buffer& writeBuffer)
+{
+    SocketOperation op = _decoratee->initialize(readBuffer, writeBuffer);
+
+    if (op == SocketOperationNone) // connected
+    {
+        rescheduleHeartbeat();
+        if (_enableIdleCheck)
+        {
+            // Reschedule because with SSL, the connection is connected after a read.
+            _timer->schedule(_idleCheckTimerTask, _idleTimeout, true);
+        }
+    }
+
+    return op;
+}
+
+IdleTimeoutTransceiverDecorator::~IdleTimeoutTransceiverDecorator()
+{
+    // If we destroy this object before calling init(), _heartbeatTimerTask and _idleCheckTimerTask will be null.
+    if (_heartbeatTimerTask)
+    {
+        _timer->cancel(_heartbeatTimerTask);
+    }
+    if (_idleCheckTimerTask)
+    {
+        _timer->cancel(_idleCheckTimerTask);
+    }
+}
+
+void
+IdleTimeoutTransceiverDecorator::close()
+{
+    _timer->cancel(_heartbeatTimerTask);
+    if (_enableIdleCheck)
+    {
+        _timer->cancel(_idleCheckTimerTask);
+    }
+    _decoratee->close();
+}
+
+SocketOperation
+IdleTimeoutTransceiverDecorator::write(Buffer& buf)
+{
+    // We're about to write something (maybe asynchronously) - we reschedule the heartbeat as sending a heartbeat now
+    // or soon would be redundant.
+    rescheduleHeartbeat();
```

Comment:
The whole point here is the *re* scheduling.
We're about to write something, and are several possibilities:
 a) the write succeeds synchronously
 b) the write writes one or more bytes synchronously; the remaining bytes will be written later typically because the local buffer is full
 c) the write doesn't write anything because the local buffer is full; all the bytes will be written later
 d) the write fails

In none of these cases there is any benefit in sending a concurrent or a slightly delayed heartbeat. We are very much trying to write and a heartbeat is purely about writing when there is no other write activity.

Reply:
I updated to logic to cancel the timer task  _before_ the write (for the reason described above) and then schedule it after a successful write.

Reply:
>  We are very much trying to write and a heartbeat is purely about writing when there is no other write activity.

Right but why schedule a heartbeat when writing data is in progress (b or c)? I did expect write to be implemented as follow:
```cpp

SocketOperation
IdleTimeoutTransceiverDecorator::write(Buffer& buf)
{

    _timer->cancel(_heartbeatTimerTask);
    SocketOperation op = _decoratee->write(buf);
    if (op == SocketOperationNone) // or buf.i == buf.b.end()
    {
        // We're done sending data, schedule the sending of a heartbeat.
        _timer->schedule(_heartbeatTimerTask, _idleTimeout / 2);
    }
    return op;
}

```